### PR TITLE
Seqio qualityio

### PR DESCRIPTION
**Pull request by mdehoon**
This pull request allows `Bio.SeqIO.QualityIO` to open files by file name. Using iterators instead of `readline` simplified the code, and is marginally faster.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Bio/SeqIO/QualityIO.py [hunk 1]
```diff
@@ -941,18 +941,15 @@ def FastqGeneralIterator(source):
...
                 else:
                     raise ValueError("Unexpected end of file")
-            if not seq_string.isalpha():
-                if seq_string:
-                    raise ValueError("Sequence should consists of alphabetical characters")
+            if " " in seq_string or "\t" in seq_string:
+                raise ValueError("Sequence should not contain whitespace.")
+            elif "@" in seq_string:
+                raise ValueError("Sequence should not contain '@'.")
+            seq_len = len(seq_string)
```
Comment (peterjc) [inline Bio/SeqIO/QualityIO.py]: I'm not entirely sure about relaxing the parser to allow non alphabetical characters... personally I would much prefer an error trying to parse such a file.

Comment (mdehoon) [reply Bio/SeqIO/QualityIO.py]: The commit history is confusing here. I added the `isalpha`, then I removed it again when it turned out that it doesn't do what I thought it does. Check this out:
```python
>>> "あいうえお".isalpha()
True
```
This pull request actually makes the parser _more_ strict, as previously we didn't check for `"@"`.

Comment (peterjc) [reply Bio/SeqIO/QualityIO.py]: Fair point, is-alpha wasn't as strict as I thought - I mean is-alpha and is-ascii combined? But it did catch the at sign:

```
>>> "@".isalpha()
False
```

Comment (mdehoon) [reply Bio/SeqIO/QualityIO.py]: maq at least seems to allow `.` and `~`:
http://maq.sourceforge.net/fastq.shtml

**Comment by mdehoon** [comment]
@peterjc  OK to merge?

File: Bio/SeqIO/QualityIO.py (hunk 1 above)
Comment (peterjc) [reply Bio/SeqIO/QualityIO.py]: OK, so compared to master, there was no check on the sequence contents at all (other than if a line started with "+").

You are adding a reasonable although new restriction of no "@" in the sequence. I'm OK with that, but would like it much more explicit in the commit history.

The cleanest way to do this is rebase the branch and edit the relevant commit's comment.

The easiest way to do that is we squash-merge the pull request, and write this in the combined commit's message.

**Comment by peterjc** [comment]
Oh, and there's a merge conflict now. Do you want me to sort that & adjust the commit messages at the same time?

**Comment by mdehoon** [comment]
@peterjc 

> Oh, and there's a merge conflict now. Do you want me to sort that & adjust the commit messages at the same time?

Yes that would be great, thanks!

File: Tests/test_SeqIO_QualityIO.py [hunk 2]
```diff
@@ -181,7 +181,7 @@ def check_qual_char(self, filename, good_count, count):
          ("long_qual", 3),
          ("short_qual", 2),
          ("double_seq", 3),
-         ("double_qual", 2),
+         ("double_qual", 3),
```
Comment (peterjc) [inline Tests/test_SeqIO_QualityIO.py]: This is a functional change  - I think the example contains two good records, then the third record is malformed.

Yes if you only look at the first four lines the third record is OK - but that's exactly the point of the test case, it has a repeated quality block making it malformed.

File: Tests/Quality/error_qual_space.fastq [hunk 3]
```diff
@@ -18,4 +18,3 @@ YYYYYYYYYYYYYYYYYY WWWYWY
 GTATTATTTAATGGCATACACTCAA
 +SLXA-B3_649_FC8437_R1_1_1_183_714
 YYYYYYYYYYWYYYYWYWWUWWWQQ
-
```
Comment (peterjc) [inline Tests/Quality/error_qual_space.fastq]: True the blank line isn't needed, and in hindsight ought not to be there, but this is the file as included in the paper's supplementary materials so I would prefer not to change it: 
https://academic.oup.com/nar/article/38/6/1767/3112533

File: Tests/Quality/error_qual_tab.fastq [hunk 4]
```diff
@@ -18,4 +18,3 @@ YYYYYYYYYYYYYYYYYYWWWWYWY
 GTATTATTTAATGGCATACACTCAA
 +SLXA-B3_649_FC8437_R1_1_1_183_714
 YYYYYYYYYY	YYYYWYWWUWWWQQ
-
```
Comment (peterjc) [inline Tests/Quality/error_qual_tab.fastq]: Again the blank line isn't needed, and in hindsight ought not to be there, but this is the file as included in the paper's supplementary materials so I would prefer not to change it: 
https://academic.oup.com/nar/article/38/6/1767/3112533

**Comment by peterjc** [comment]
Shall I pull out the bits I have no objection to as a separate branch so we can get them merged today?

And then rebase this branch to contain just the changes I am querying?

**Comment by mdehoon** [comment]
@peterjc  OK.

**Comment by mdehoon** [comment]
@peterjc  Just to confirm, are you waiting for me or should I wait for you? Just to make sure we are not waiting for each other.

**Comment by peterjc** [comment]
Waiting on me finding a quiet moment to tackle the git stuff needed...

**Comment by mdehoon** [comment]
@peterjc  OK, no problem. Thanks.

**Comment by peterjc** [comment]
It doesn't look too hard to resolve the conflicts now that #2744 is merged, but it isn't trivial.

**Comment by mdehoon** [comment]
It's easier to handle the remaining changes separately; closing this PR.

**Pull request closed without merging**